decode/zlibg2gini.1
." $Id: zlibg2gini.1,v 1.0 2004/09/16 10:00:00 yoksas Exp $
.TH ZLIBG2GINI 1 "2004/09/16 10:00:00 $"
.SH NAME
zlibg2gini \- Unidata LDM decoder for ZLIB-compressed NOAAPORT GINI imagery
.SH SYNOPSIS
.HP
.ft B
zlibg2gini
.nh
\%[-f\ \fIZLIBfile\fP]
\%[-v]
\%[-x]
\%[-l\ \fIlogpath\fP]
\%\fIIMGfile\fP
.hy
.ft
.SH DESCRIPTION
.LP
This program is an LDM decoder for NOAAPORT image data of type \fBGINI\fP
that has been ZLIB-compressed.
It reads a ZLIB-compressed GINI data product from standard input or from a file
and decodes that data into a NOAAPORT format GINI file.
.LP
This program can be invoked by an entry in the LDM \fBpqact.conf\fP
configuration file with a pattern similar to the following:
.sp
.nf
#
# NOAAPORT ZLIB-compressed images decoded into GINI files
#
FNEXRAD	^satz/ch[0-9]/.*/(.*)/([12][0-9])([0-9][0-9])([01][0-9])([0-3][0-9]) ([0-2][0-9])([0-5][0-9])/(.*)/(.*km)/
	PIPE	-close
	zlibg2gini -vl logs/ldm-mcidas.log
	/data/ldm/gempak/images/sat/\\8/\\9/\\1/\\1_\\2\\3\\4\\5_\\6\\7

.fi
.RE
.LP
Since Unidata-supported display and analysis applications (GEMPAK, the IDF,
and McIDAS) can read the ZLIB-compressed images directly, use of this decoder
should be infrequent.
.fi
.RE
.LP
This program can also be invoked as a
stand-alone program, outside the LDM environment.

.SH FLAGS
.TP
.BI "-f " ZLIBfile
Pathname for input ZLIB compressed GINI file (default is stdin)
.TP
.B "-h "
Print help information, then exit the program.
.TP
.B -v
Verbose logging.
Informative messages (level LOG_INFO) are logged.
By default, only messages of severity LOG_NOTICE and greater are
logged. See the Unidata  LDM Site  Manager's  Guide  for more information on 
message logging.
.TP
.B -x
Debug logging.
Debugging messages (level LOG_DEBUG) are logged. See the Unidata  LDM Site  
Manager's  Guide  for more information on message logging.
.TP
.BI "-l " logpath
Log file pathname.
The program uses Unidata's \fBulog\fP(3) package to write error and diagnostic
messages.
The default is to write to the \fBsyslogd\fP(8) daemon.  If this option is
specified, however, then messages will be written to the file
\fIlogpath\fP unless that pathname is "-", in which case all messages will be
written to standard error.

.RE
.SH NOTES
.LP
.RE
.SH "SEE ALSO"
.LP
.BR ~ldm-mcidas/etc/ldm-mcidas-pqact.conf,
.BR ldmd (1),
.BR ulog (3),
.BR syslogd (8).


RELEASE_NOTES
VERSION	COMMENTS
-------	-------------------------------------------------------------

2012    2012-01-18

        Updated version number to 2012 to reflect small updates (e.g.,
        new src/port/cfortran/cfortran.h, modified src/decode/Makefile)
        and that the package is still viable.

2008    2009-04-30

        Note: The version number is kept at 2008 since McIDAS went
        into maintenance mode with v2008.

        A variety of changes to C routines, Makefiles and test data
        files.

        Introduction of a new McIDAS lightning decoder, 'lgt2md', that
        handles both NLDN binary and USPLN ASCII data record formats.
        'lgt2md' is designed as a drop-in replacement for 'nldn2md':
        user's pattern action file entries for decoding NLDN data
        can be easily updated by replacing 'nldn2md' with 'lgt2md'.

        ROUTE.SYS (distributed as route.sys) was updated to include
        a new product, 'LP', for USPLN lightning data.

        The Makefile in the src/decode directory was reorganized to
        build the subset of the encoders/decoders contained in the package
        that are written entirely in C:

        area2png    - PNG-compresses the image portion of single-banded
                      McIDAS AREA files
        pnga2area   - uncompresses images created by 'area2png'
        zlibg2gini  - uncompresses the Zlib-compressed images contained
                      in the NOAAPort broadcast (and conveyed in the
                      IDD in the NIMAGE feed)
        pngg2gini   - uncompress NOAAPort images that have been
                      PNG-compressed by the Unidata NOAAPort ingest
                      package (archaic)
        proftomd    - decodes wind profiler netCDF files from NOAA/GSD
                      into McIDAS MD format files
        lgt2md      - decodes NLDN and USPLN lightning data into
                      McIDAS MD files

        This set is built by running 'make' or 'make basic'.
        
        In addition to this set, the Fortran-based 'nldn2md' NLDN
        lightning decoder will be built by 'make full'.

        The Makefile in src/decode also provides for the (attempted)
        building of archaic McIDAS decoders:

        cdftomd     - decodes wind profiler netCDF files from NOAA/GSD
                      into McIDAS MD format files (use 'proftomd' instead)
        gunrv2      - decodes compressed McIDAS GRID files sent in the
                      old Zephyr/Alden satellite broadcast into McIDAS
                      GRID format
        lwfile
        lwtmd2      - decodes compressed McIDAS MD files sent in the
                      old Zephyr/Alden satellite broadcast into McIDAS
                      MD format
        lwtoa3      - decodes delta-compressed McIDAS image files from
                      the old Zephyr/Alden satellite broadcast into McIDAS
                      AREA format
        nids2area   - converts Nexrad Level III images into McIDAS
                      AREA file format (best to use McIDAS IMGCOPY from
                      NEXR to AREA datasets directly as McIDAS capabilities
                      are kept up to date)
 
        In addition decoders built by 'make full', an attepmt will be
        made to build these decoders when the user runs 'make all'.
        There is _no_ guarantee that these decoders will be built
        successfully!

2004    2004-09-14

        Added O_TRUNC flag to mkdirs_open call in pnga2area.c,
        pngg2gini.c, and zlibg2gini.c.  This corrects a problem
        encountered when decoding Unidata-Wisconsin images (IDD
        feedtype UNIWISC aka MCIDAS) when using McIDAS routing table
        (ROUTE.SYS) concepts.  The problem was manifested in decoded
        image files keeping the size of the images they were
        overwriting.  So, the file size was always the "high water
        mark" of sizes for the products.

        NOTES:
        - this did not affect the sizes of files decoded without using
          McIDAS routing table concepts
        - the decoded images using McIDAS routing table concepts were
          valid/usable even though they are larger than they need to
          be

        Changed IDD datastream name from MCIDAS to UNIWISC in example
        pqact.conf files.

        Updated distribution with the versions of SATANNOT, SATBAND,
        and SCHEMA from the latest Unidata McIDAS distribution, v2004.

        Cleaned up batch.k shell script that runs McIDAS postprocess
        BATCH files.  Internal comments should now be easier to follow
        than in previous versions.

        Updated man pages for currently used decoders: area2png(1),
        nldn2md(1), pnga2area(1), pngg2gini(1), proftomd(1), and
        zlibg2gini(1).

        Removed building of decoders now considered to be obsolete:
        cdftomd(1), lwtoa3(1), gunrv2(1), lwfile(1), lwtmd2(1),
        and nids2area(1).  Source code for these decoders is still
        included in the distribution, but their use is not guaranteed
        mainly since subroutines they use from the McIDAS library
        are no longer supported.

2003    2003-07-29

        Added utililty routine zlibg2gini that uncompresses Zlib
        compressed GINI images in NIMAGE IDD datastream.

        Fixed a bug that kept McIDAS routing table, ROUTE.SYS,
        dates and times from being updated.

2002    2002-12-20

        Replaced information messages in pngg2gini.c that referred to
        pnga2area with pngg2gini.

        Corrected logic in pnga2area.c for processing a product whose
        product code is not contained in the McIDAS routing table,
        ROUTE.SYS.  Before, the routing table would get corrupted
        and the AREA file number included in the '-r pcode,anum'
        flag would not be used.  Now, the routing table will not
        be updated for unregistered products, and the default AREA
        file number will be used.
	

decode/zlibg2gini.c
#define UD_FORTRAN_SOURCE

#include <unistd.h>
#include <stdlib.h>
#include <stdio.h>
#include <assert.h>
#include <signal.h>
#include <unistd.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <fcntl.h>
#include <limits.h>
#ifndef PATH_MAX
#  define PATH_MAX 255
#endif
#include "zlib.h"
#include "zutil.h"
#include "mkdirs_open.h"
#include "ulog.h"
#include "alarm.h"

#define GINI_HED_LEN 533
#define NUMREAD      10000
#define WMO_HED_LEN  21
#define ZLIB_BUF_LEN 10000

/*
** Some needed function prototypes <<<<< UPC add 20080918 >>>>>
*/
int getbuf( int, char *, int, int );
int GetInt( unsigned char *, int );

/*
** The following are needed since this links against imgsubs.o which needs
** to have these defined.
*/

char            *annotfile=NULL;
char            *bandfile=NULL;

char             mcpathdir[PATH_MAX];                /* McIDAS MCPATH string */
char             pathname[PATH_MAX];        /* fully qualified file pathname */


/*
** On to declarations that are used
*/

unsigned         alarm_timeout = 60;              /* seconds; should suffice */


/*
** Function prototypes
*/

char      *GiniFileName( unsigned char *, char * );
void       eat_stream_and_exit( int );

/*
 * SIGALRM handler.
 */
static void
alarm_handler( int sig )
{
    assert( SIGALRM == sig );
    uerror( "zlibggini: no data read in %u seconds, exiting", alarm_timeout );
    exit( 1 );
}

/*
** Called at exit.
** This callback routine registered by atexit().
*/

static void
cleanup()
{
    (void) closeulog();
}

/*
** Called upon receipt of signals.
** This callback routine registered in set_sigactions() .
*/

static void
signal_handler( int sig )
{
    switch( sig ) {
      case SIGHUP:
                   udebug( "SIGHUP" ) ;
                   return ;
      case SIGINT:
                   unotice( "Interrupt" ) ;
                   exit( 0 );
      case SIGTERM:
                   udebug( "SIGTERM" ) ;
                   exit( 0 );
      case SIGUSR1:
                   udebug( "SIGUSR1" ) ;
                   return;
      case SIGUSR2:
                   if ( toggleulogpri(LOG_INFO) )
                     unotice( "Going verbose" ) ;
                   else
                     unotice( "Going silent" ) ;
                   return;
      case SIGPIPE:
                   unotice( "SIGPIPE" ) ;
                   exit( 0 );
    }

    udebug( "signal_handler: unhandled signal: %d", sig );
}

static void
set_sigactions()
{
    struct sigaction sigact ;

    sigact.sa_handler   = signal_handler;
    sigemptyset(&sigact.sa_mask) ;
    sigact.sa_flags     = 0 ;

    (void) sigaction( SIGHUP,  &sigact, (struct sigaction *) NULL );
    (void) sigaction( SIGINT,  &sigact, (struct sigaction *) NULL );
    (void) sigaction( SIGTERM, &sigact, (struct sigaction *) NULL );
    (void) sigaction( SIGUSR1, &sigact, (struct sigaction *) NULL );
    (void) sigaction( SIGUSR2, &sigact, (struct sigaction *) NULL );
    (void) sigaction( SIGPIPE, &sigact, (struct sigaction *) NULL );
}


/*
** Usage
*/

static void usage( char *argv0 )
{
#define USAGE_FMT "\
Example Usage:\n\
       %s <-vxl-> IMGfile < infile \n\
       %s <-vxl-> -f infile IMGfile\n\
Flags:\n\
    Help:\n\
    -h              Print the help, then exit the program\n\n\
    Satellite definitions:\n\
    -f infile       Pathname for input PNG compressed file (default is stdin)\n\n\
    Logging:\n\
    -l logname      Log file name (default uses local0; use '-' for stdout)\n\
    -v              Verbose logging\n\
    -x              Debug logging\n\n\
Positionals:\n\
    IMGfile         Output image pathname (no default)\n\n\
Example Invocation:\n\
    %s -f zlibfile outfile\n\n\
    -- or --\n\n\
    %s outfile < zlibfile\n\n"

    (void) fprintf (stderr, USAGE_FMT, argv0, argv0, argv0, argv0);
}

static char *logfname = "";

int
main( int argc, char *argv[] )
{

    Bytef          uncompr[ZLIB_BUF_LEN];
    uLong          uncomprLen=ZLIB_BUF_LEN;
    
    unsigned long  intot;
    unsigned long  outtot=0;

    char          *infname  = NULL;
    char          *outfname = NULL;

    char           buf[101]={0};

    const char    *c;

    unsigned char *b;
    unsigned char *p;

    unsigned char  imghed[GINI_HED_LEN+1];
    unsigned char  data[NUMREAD];

    int            ch;
    int            hoff=0;
    int            numin;
    int            rc;

    int            elems;
    int            lines;
    int            nline;
    int            totlin = 0;

    int            logmask = (LOG_MASK(LOG_ERR) | LOG_MASK(LOG_NOTICE)) ;
    int            logfd;

    size_t         nbyte;
    size_t         nmove;

    z_stream       d_stream;                    /* Zlib decompression stream */

    mode_t         mode = (S_IRUSR | S_IWUSR | S_IRGRP | S_IROTH);
    int            fd_in = 0;                              /* standard input */
    int            fd_out;

    extern int    optind;
    extern int    opterr;
    extern char  *optarg;

    int           alarm_initialized = 0;
    Alarm         timeout_alarm;


    /*
    ** If nothing was specified on the command line, print usage and exit.
    */
    if ( argc == 1 ) {
      usage( argv[0] );
      exit( 0 );
    }

    opterr = 1;

    while ( (ch = getopt(argc, argv, "hf:vxl:")) != EOF )
      switch (ch) {
        case 'h':
        case '?':
                 usage( argv[0] );
                 exit( 0 );
                 break;
        case 'f':
                 infname = optarg;
                 break;
        case 'v':
                 logmask |= LOG_MASK(LOG_INFO);
                 break;
        case 'x':
                 logmask |= LOG_MASK(LOG_DEBUG);
                 break;
        case 'l':
                 if ( optarg[0] == '-' && optarg[1] != 0 ) {
                   fprintf( stderr, "logfile \"%s\" ??\n", optarg );
                   usage( argv[0] );
                   exit( 0 );
                 }
                 /* else */
                 logfname = optarg;
                 break;
      }

    /*
    ** Sanity checks
    */

    if ( argc - optind < 0 ) {
      usage( argv[0] );
      exit( 0 );
    }

    if ( argc != (optind + 1) ) {          /* output file name not specified */
      printf( "No output file name specified, exiting\n" );
      exit( 1 );
    }


    /*
    ** If user specified an input file with '-f' flag, redirect
    ** stdin to the file aborting on error.
    */

    if ( infname != (char *) NULL ) {
      if ( freopen( infname, "r", stdin ) == (FILE *) NULL ) {
        printf( "Can't open input file: %s\n", infname );
        exit( 1 );
      }
    }


    /*
    ** Setup syslog logging.
    */

    (void) setulogmask( logmask );

    if ( logfname == NULL || !(*logfname == '-' && logfname[1] == 0) )
      (void) fclose( stderr );

    logfd = openulog( ubasename(argv[0]), (LOG_CONS|LOG_PID), LOG_LDM,
                      logfname );

    if ( logfname == NULL || !(*logfname == '-' && logfname[1] == 0) ) {
      setbuf( fdopen(logfd, "a" ), NULL);
    }


    /*
    ** Start
    */

    unotice( "Starting Up" );

    /*
    ** Register exit handler, establish signal handlers, and set alarm handler.
    */

    (void) atexit( cleanup );

    set_sigactions();

    if ( !alarm_initialized ) {
      alarm_init( &timeout_alarm, alarm_timeout, alarm_handler );
      alarm_initialized = 1;
    }

    alarm_on( &timeout_alarm );

    /*
    ** Actual processing
    */

    rc = getbuf( fd_in, (char *)data, 1, (size_t) NUMREAD );
    (void) memcpy( buf, data, (size_t)100 );
    buf[100] = 0;

    p = (unsigned char *) strstr( (const char *)buf, "\r\r\n" );
    if ( p == (unsigned char *) NULL ) {
      uerror( "header not found in compressed image" );
      eat_stream_and_exit( 1 );
    }

    intot = p + 3 - (unsigned char *) buf;
    b     = data + intot;
    numin = NUMREAD - intot;

    /*
    ** Get the Zlib compressed image header
    */

    d_stream.zalloc    = (alloc_func) 0;
    d_stream.zfree     = (free_func) 0;
    d_stream.opaque    = (voidpf) 0;
    d_stream.next_in   = (Bytef *) b;
    d_stream.avail_in  = (uInt) numin;
    d_stream.next_out  = uncompr;
    d_stream.avail_out = (uInt) uncomprLen;

    rc = inflateInit( &d_stream );
    if ( rc != Z_OK ) {
      uerror( "inflateInit error, exiting!" );
      eat_stream_and_exit( 2 );
    }

    rc = inflate( &d_stream, Z_NO_FLUSH );
    if ( rc != Z_OK && rc != Z_STREAM_END ) {
      fprintf( stderr, "inflate error, exiting!\n" );
      eat_stream_and_exit( 3 );
    }

    if ( d_stream.total_out != GINI_HED_LEN ) {
      fprintf( stderr, "inflated image header size error, exiting!\n" );
      eat_stream_and_exit( 4 );
    }

    (void) memcpy( imghed, uncompr, (size_t)GINI_HED_LEN );
    imghed[GINI_HED_LEN] = 0;

    if ( imghed[21] != 1 ) {
      uerror( "Input fails source test: 1 != %c", imghed[21] );
      eat_stream_and_exit( 5 );
    }

    if ( imghed[63] != 0 ) {                /* zero out the compression flag */
      imghed[63] = 0;
    }

    c = strstr( (const char *)imghed, "KNES" );
    if ( c != (const char *) NULL ) {                       /* 'KNES' found  */
      c = strstr( c, "\n" );
      if ( c != (const char *) NULL ) {                     /* NL found      */
        hoff = c - (const char *) imghed + 1;
      }
    }

    if ( hoff != WMO_HED_LEN ) {
      uerror( "image header length mismatch" );
      eat_stream_and_exit( 6 );
    }

    elems = GetInt( (unsigned char *)imghed + hoff + 16, 2 );
    lines = GetInt( (unsigned char *)imghed + hoff + 18, 2 );

    udebug( "Input image:: lines = %d, elements = %d", lines, elems );

    /*
    ** Create the output file name:
    **
    **   IMGfile  - file name or mask specified.  In this case, we need to
    **              expand the "replaceables" (\BAND, \RES, and \SAT) and
    **              strftime format specifications (e.g., %Y, %m, %d, etc.)
    **              in user specified filename using information from the
    **              GINI file header.
    */

    outfname = GiniFileName( imghed, argv[optind] );
    if ( outfname == (char *) NULL ) {
      uerror( "Error creating output name from pattern: %s", argv[optind] );
      eat_stream_and_exit( 7 );
    }

    uinfo( "output file pathname: %s", outfname );

    /* <<<<< UPC mod 20040914 - add file truncation >>>>> */
    fd_out = mkdirs_open( outfname, O_CREAT|O_WRONLY|O_TRUNC, mode );
    if ( fd_out == -1 ) {             /* make sure directories to file exist */
      uerror( "Can't open file %s", outfname );
      eat_stream_and_exit( 8 );
    }


    do {

      intot  += d_stream.total_in;
      outtot += d_stream.total_out;

      nline   = d_stream.total_out / elems;
      totlin += nline;

      if ( totlin > lines ) break;

      write( fd_out, uncompr, d_stream.total_out );

      udebug( "bytes in: %4d, bytes out: %4d, nlines = %2d, total lines = %4d",
               d_stream.total_in, d_stream.total_out, nline, totlin );

      nmove = numin - d_stream.total_in;
      if ( nmove > 0 )
        (void) memmove( data, b+d_stream.total_in, nmove );
      else
        break;

      nbyte = NUMREAD - nmove;
      numin = nmove + getbuf( fd_in, (char *)(data+nmove), 1, nbyte );
      b     = data;

      d_stream.zalloc    = (alloc_func) 0;
      d_stream.zfree     = (free_func) 0;
      d_stream.opaque    = (voidpf) 0;
      d_stream.next_in   = (Bytef *) b;
      d_stream.avail_in  = (uInt) numin;
      d_stream.next_out  = uncompr;
      d_stream.avail_out = (uInt) uncomprLen;

      rc = inflateInit( &d_stream );
      if ( rc != Z_OK ) {
        uerror( "inflateInit error, exiting!" );
        break;
      }

      rc = inflate( &d_stream, Z_NO_FLUSH );

    } while( (rc == Z_STREAM_END) || (rc == Z_OK) );


    alarm_off( &timeout_alarm );

    /*
    ** Close the output file and exit.
    */

    close( fd_out );
    fclose( stdin );

    unotice( "unZlib:: %8ld  %8ld  %6.4f",
             intot, outtot, ((float)outtot / intot) );

    unotice( "Exiting" );

    exit( 0 );

}
